File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/data/offline/OfflineMediaStore.kt
package mx.nexara.mobile.nativeapp.data.offline

import android.content.Context
import android.util.Base64
import java.io.File
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * Persiste blobs de media (data URLs) en disco para no inflar la cola JSON offline.
 * Encola referencias `nexara-media://{id}` y las reexpande al hacer replay.
 */
class OfflineMediaStore(context: Context) {
    private val dir = File(context.filesDir, "nexara_offline_media").also { it.mkdirs() }
    private val mimeById = ConcurrentHashMap<String, String>()

    fun externalizeDataUrls(body: String?): String? {
        if (body.isNullOrBlank()) return body
        if (!body.contains("data:")) return body
        val matcher = DATA_URL.matcher(body)
        val sb = StringBuffer()
        while (matcher.find()) {
            val mime = matcher.group(1) ?: "application/octet-stream"
            val b64 = matcher.group(2) ?: continue
            val id = saveBase64(b64, mime) ?: continue
            matcher.appendReplacement(sb, Matcher.quoteReplacement("nexara-media://$id"))
        }
        matcher.appendTail(sb)
        return sb.toString()
    }

    fun expandMediaRefs(body: String?): String? {
        if (body.isNullOrBlank()) return body
        if (!body.contains("nexara-media://")) return body
        val matcher = MEDIA_REF.matcher(body)
        val sb = StringBuffer()
        while (matcher.find()) {
            val id = matcher.group(1) ?: continue
            val dataUrl = loadAsDataUrl(id) ?: continue
            matcher.appendReplacement(sb, Matcher.quoteReplacement(dataUrl))
        }
        matcher.appendTail(sb)
        return sb.toString()
    }

    fun delete(id: String) {
        File(dir, id).delete()
        mimeById.remove(id)
        File(dir, "$id.mime").delete()
    }

    fun purgeRefsInBody(body: String?) {
        if (body.isNullOrBlank()) return
        val matcher = MEDIA_REF.matcher(body)
        while (matcher.find()) {
            matcher.group(1)?.let { delete(it) }
        }
    }

    private fun saveBase64(b64: String, mime: String): String? {
        return try {
            val bytes = Base64.decode(b64, Base64.DEFAULT)
            val id = UUID.randomUUID().toString()
            File(dir, id).writeBytes(bytes)
            File(dir, "$id.mime").writeText(mime)
            mimeById[id] = mime
            id
        } catch (_: Exception) {
            null
        }
    }

    private fun loadAsDataUrl(id: String): String? {
        val file = File(dir, id)
        if (!file.exists()) return null
        return try {
            val mime = mimeById[id]
                ?: File(dir, "$id.mime").takeIf { it.exists() }?.readText()?.trim()
                ?: "application/octet-stream"
            val b64 = Base64.encodeToString(file.readBytes(), Base64.NO_WRAP)
            "data:$mime;base64,$b64"
        } catch (_: Exception) {
            null
        }
    }

    companion object {
        private val DATA_URL: Pattern =
            Pattern.compile("data:([\\w/+.-]+);base64,([A-Za-z0-9+/=\\r\\n]+)")
        private val MEDIA_REF: Pattern =
            Pattern.compile("nexara-media://([0-9a-fA-F\\-]{36})")
    }
}

File: apps/mobile-native/ios/NexaraApp/Offline/OfflineMediaStore.swift
import Foundation

/// Persiste blobs de media (data URLs) en disco para no inflar la cola JSON offline.
/// Usa referencias `nexara-media://{id}` en el body encolado y las reexpande al replay.
final class OfflineMediaStore {
    static let shared = OfflineMediaStore()

    private let dir: URL
    private let queue = DispatchQueue(label: "mx.nexara.offline.media")

    private init() {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        dir = base.appendingPathComponent("nexara-offline-media", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    func externalizeDataUrls(_ body: String?) -> String? {
        guard var body, body.contains("data:") else { return body }
        let pattern = #"data:([\w/+.-]+);base64,([A-Za-z0-9+/=\r\n]+)"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return body }
        let ns = body as NSString
        let matches = regex.matches(in: body, range: NSRange(location: 0, length: ns.length))
        for match in matches.reversed() {
            guard match.numberOfRanges >= 3,
                  let mimeRange = Range(match.range(at: 1), in: body),
                  let b64Range = Range(match.range(at: 2), in: body),
                  let fullRange = Range(match.range, in: body)
            else { continue }
            let mime = String(body[mimeRange])
            let b64 = String(body[b64Range])
            guard let id = saveBase64(b64, mime: mime) else { continue }
            body.replaceSubrange(fullRange, with: "nexara-media://\(id)")
        }
        return body
    }

    func expandMediaRefs(_ body: String?) -> String? {
        guard var body, body.contains("nexara-media://") else { return body }
        let pattern = #"nexara-media://([0-9a-fA-F\-]{36})"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return body }
        let ns = body as NSString
        let matches = regex.matches(in: body, range: NSRange(location: 0, length: ns.length))
        for match in matches.reversed() {
            guard match.numberOfRanges >= 2,
                  let idRange = Range(match.range(at: 1), in: body),
                  let fullRange = Range(match.range, in: body)
            else { continue }
            let id = String(body[idRange])
            guard let dataUrl = loadAsDataUrl(id) else { continue }
            body.replaceSubrange(fullRange, with: dataUrl)
        }
        return body
    }

    func purgeRefs(in body: String?) {
        guard let body, body.contains("nexara-media://") else { return }
        let pattern = #"nexara-media://([0-9a-fA-F\-]{36})"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return }
        let ns = body as NSString
        for match in regex.matches(in: body, range: NSRange(location: 0, length: ns.length)) {
            guard match.numberOfRanges >= 2,
                  let idRange = Range(match.range(at: 1), in: body)
            else { continue }
            delete(String(body[idRange]))
        }
    }

    private func saveBase64(_ b64: String, mime: String) -> String? {
        queue.sync {
            guard let data = Data(base64Encoded: b64, options: [.ignoreUnknownCharacters]) else { return nil }
            let id = UUID().uuidString
            let file = dir.appendingPathComponent(id)
            let mimeFile = dir.appendingPathComponent("\(id).mime")
            do {
                try data.write(to: file, options: .atomic)
                try mime.data(using: .utf8)?.write(to: mimeFile, options: .atomic)
                return id
            } catch {
                return nil
            }
        }
    }

    private func loadAsDataUrl(_ id: String) -> String? {
        queue.sync {
            let file = dir.appendingPathComponent(id)
            guard let data = try? Data(contentsOf: file) else { return nil }
            let mime = (try? String(contentsOf: dir.appendingPathComponent("\(id).mime"), encoding: .utf8))?
                .trimmingCharacters(in: .whitespacesAndNewlines)
                ?? "application/octet-stream"
            return "data:\(mime);base64,\(data.base64EncodedString())"
        }
    }

    private func delete(_ id: String) {
        queue.sync {
            try? FileManager.default.removeItem(at: dir.appendingPathComponent(id))
            try? FileManager.default.removeItem(at: dir.appendingPathComponent("\(id).mime"))
        }
    }
}
